File: src/core/SqlFormatter.cpp
#include "SqlFormatter.h"
#include <regex>
#include <sstream>
#include <map>
#include <stdexcept>
#include <cctype>

namespace SqlFormatter {

std::string formatSql(const std::string& sql) {
    if (sql.empty()) {
        return "Not found";
    }
    
    std::string formatted = sql;
    std::vector<std::string> keywords = {
        "SELECT", "FROM", "WHERE", "AND", "OR", "ORDER BY", "GROUP BY"
    };
    
    for (const auto& keyword : keywords) {
        std::string pattern = " " + keyword + " ";
        std::string replacement = "\n" + keyword + " ";
        
        size_t pos = 0;
        // Case-insensitive search
        std::string upperFormatted = formatted;
        std::transform(upperFormatted.begin(), upperFormatted.end(), 
                       upperFormatted.begin(), ::toupper);
        
        while ((pos = upperFormatted.find(pattern, pos)) != std::string::npos) {
            formatted.replace(pos, pattern.length(), replacement);
            upperFormatted = formatted;
            std::transform(upperFormatted.begin(), upperFormatted.end(), 
                           upperFormatted.begin(), ::toupper);
            pos += replacement.length();
        }
    }
    
    // Trim leading/trailing whitespace
    size_t start = formatted.find_first_not_of(" \t\n\r");
    size_t end = formatted.find_last_not_of(" \t\n\r");
    
    if (start == std::string::npos) return "";
    return formatted.substr(start, end - start + 1);
}

std::string formatParams(const std::vector<std::string>& params) {
    if (params.empty()) {
        return "Not found";
    }
    
    std::stringstream ss;
    for (const auto& param : params) {
        // Parse format: TYPE:INDEX:VALUE
        size_t firstColon = param.find(':');
        size_t secondColon = param.find(':', firstColon + 1);
        
        if (firstColon != std::string::npos && secondColon != std::string::npos) {
            std::string paramType = param.substr(0, firstColon);
            std::string paramIndex = param.substr(firstColon + 1, 
                                                   secondColon - firstColon - 1);
            std::string paramValue = param.substr(secondColon + 1);
            
            ss << "  [" << paramIndex << "] " << paramType << ": " << paramValue << "\n";
        } else {
            ss << "  " << param << "\n";
        }
    }
    
    return ss.str();
}

std::string replacePlaceholders(const std::string& query, 
                                 const std::vector<std::string>& params) {
    // Build map of position -> value
    std::map<int, std::string> valuesByPos;
    
    for (const auto& param : params) {
        // Parse format: TYPE:INDEX:VALUE
        size_t firstColon = param.find(':');
        size_t secondColon = param.find(':', firstColon + 1);
        
        if (firstColon == std::string::npos || secondColon == std::string::npos) {
            continue;
        }
        
        std::string type = param.substr(0, firstColon);
        int pos = std::stoi(param.substr(firstColon + 1, secondColon - firstColon - 1));
        std::string value = param.substr(secondColon + 1);
        
        // Convert to lowercase for comparison
        std::string typeLower = type;
        std::transform(typeLower.begin(), typeLower.end(), typeLower.begin(), ::tolower);
        
        std::string parsedValue;
        
        if (typeLower == "string") {
            // Escape single quotes for SQL
            std::string escaped;
            for (char c : value) {
                if (c == '\'') escaped += "''";
                else escaped += c;
            }
            parsedValue = "'" + escaped + "'";
        } else if (typeLower == "bigdecimal" || typeLower == "number" || 
                   typeLower == "int" || typeLower == "long" || typeLower == "float") {
            // Numeric values - use as-is
            parsedValue = value;
        } else {
            throw std::runtime_error("Unsupported type: " + type);
        }
        
        valuesByPos[pos] = parsedValue;
    }
    
    // Replace ? placeholders in order
    std::string result;
    int index = 1;
    
    for (size_t i = 0; i < query.size(); ++i) {
        if (query[i] == '?') {
            auto it = valuesByPos.find(index);
            if (it != valuesByPos.end()) {
                result += it->second;
            } else {
                throw std::runtime_error("Missing value for position " + 
                                         std::to_string(index));
            }
            ++index;
        } else {
            result += query[i];
        }
    }
    
    return result;
}

} // namespace SqlFormatter

File: src/core/sql_formatter.rs
//! SQL formatting utilities.
//!
//! Provides functions to format SQL queries with line breaks at keywords,
//! format parameter lists, and replace placeholders with actual values.

use std::collections::HashMap;

/// Format SQL with line breaks at major keywords for readability.
pub fn format_sql(sql: &str) -> String {
    if sql.is_empty() {
        return "Not found".to_string();
    }

    let keywords = ["SELECT", "FROM", "WHERE", "AND", "OR", "ORDER BY", "GROUP BY"];
    let mut formatted = sql.to_string();

    for keyword in keywords {
        let pattern = format!(" {} ", keyword);
        let replacement = format!("\n{} ", keyword);

        // Case-insensitive replacement
        let upper_formatted = formatted.to_uppercase();
        let mut result = String::new();
        let mut last_end = 0;

        for (start, _) in upper_formatted.match_indices(&pattern) {
            result.push_str(&formatted[last_end..start]);
            result.push_str(&replacement);
            last_end = start + pattern.len();
        }
        result.push_str(&formatted[last_end..]);
        formatted = result;
    }

    formatted.trim().to_string()
}

/// Format a parameter list for display.
///
/// Parses parameters in the format `TYPE:INDEX:VALUE` and formats them nicely.
pub fn format_params(params: &[String]) -> String {
    if params.is_empty() {
        return "Not found".to_string();
    }

    let mut result = String::new();

    for param in params {
        // Parse format: TYPE:INDEX:VALUE
        let parts: Vec<&str> = param.splitn(3, ':').collect();
        
        if parts.len() == 3 {
            let param_type = parts[0];
            let param_index = parts[1];
            let param_value = parts[2];
            result.push_str(&format!("  [{}] {}: {}\n", param_index, param_type, param_value));
        } else {
            result.push_str(&format!("  {}\n", param));
        }
    }

    result
}

/// Replace `?` placeholders in SQL with actual parameter values.
///
/// Parameters must be in the format `TYPE:INDEX:VALUE`.
/// String values are quoted and escaped, numeric values are used as-is.
pub fn replace_placeholders(query: &str, params: &[String]) -> Result<String, String> {
    // Build map of position -> value
    let mut values_by_pos: HashMap<i32, String> = HashMap::new();

    for param in params {
        let parts: Vec<&str> = param.splitn(3, ':').collect();
        
        if parts.len() != 3 {
            continue;
        }

        let param_type = parts[0].to_lowercase();
        let pos: i32 = parts[1].parse().map_err(|_| "Invalid parameter index")?;
        let value = parts[2];

        // Handle NULL explicitly
        if value == "null" {
            values_by_pos.insert(pos, "NULL".to_string());
            continue;
        }

        let parsed_value = match param_type.as_str() {
            "string" | "timestamp" | "date" => {
                // Escape single quotes for SQL
                let escaped = value.replace('\'', "''");
                format!("'{}'", escaped)
            }
            "boolean" => {
                 value.to_uppercase()
            }
            "bigdecimal" | "number" | "int" | "long" | "float" | "double" => {
                value.to_string()
            }
            _ => {
                // Default fallback: treat as string (safer) or error? 
                // Let's treat unknown as string to be safe against new types, or allow raw value?
                // User requirement: "Handling different parameter types correctly".
                // If it's unknown, maybe it's safest to quote it if it looks like a string?
                // Let's err on side of quoting unless it looks numeric.
                if value.chars().all(|c| c.is_numeric() || c == '.') {
                     value.to_string()
                } else {
                    let escaped = value.replace('\'', "''");
                    format!("'{}'", escaped)
                }
            }
        };

        values_by_pos.insert(pos, parsed_value);
    }

    // Replace ? placeholders in order
    let mut result = String::new();
    let mut index = 1;

    for ch in query.chars() {
        if ch == '?' {
            match values_by_pos.get(&index) {
                Some(value) => result.push_str(value),
                None => {
                    // If param missing, keep ? to indicate issue/allow debugging
                    result.push('?'); 
                    // Or return Err? User wants "debuggability". 
                    // Keeping ? with a warning might be better than failing hard?
                    // Previous logic returned Err. Let's stick to Err for strictness.
                    return Err(format!("Missing value for position {}", index));
                },
            }
            index += 1;
        } else {
            result.push(ch);
        }
    }

    Ok(result)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_sql() {
        let sql = "SELECT * FROM users WHERE id = 1 AND active = true";
        let formatted = format_sql(sql);
        assert!(formatted.contains("\nWHERE"));
        assert!(formatted.contains("\nAND"));
    }

    #[test]
    fn test_format_params() {
        let params = vec![
            "String:1:hello".to_string(),
            "Int:2:42".to_string(),
        ];
        let formatted = format_params(&params);
        assert!(formatted.contains("[1] String: hello"));
        assert!(formatted.contains("[2] Int: 42"));
    }

    #[test]
    fn test_replace_placeholders() {
        let query = "SELECT * FROM users WHERE name = ? AND id = ?";
        let params = vec![
            "String:1:John".to_string(),
            "Int:2:42".to_string(),
        ];
        let result = replace_placeholders(query, &params).unwrap();
        assert_eq!(result, "SELECT * FROM users WHERE name = 'John' AND id = 42");
    }

    #[test]
    fn test_replace_placeholders_types() {
        let query = "SELECT * FROM t WHERE a = ? AND b = ? AND c = ?";
        let params = vec![
            "String:1:null".to_string(),
            "Timestamp:2:2024-01-01 10:00:00".to_string(),
            "Boolean:3:true".to_string(),
        ];
        let result = replace_placeholders(query, &params).unwrap();
        assert_eq!(result, "SELECT * FROM t WHERE a = NULL AND b = '2024-01-01 10:00:00' AND c = TRUE");
    }

    #[test]
    fn test_escape_quotes() {
        let query = "INSERT INTO t (name) VALUES (?)";
        let params = vec!["String:1:O'Brien".to_string()];
        let result = replace_placeholders(query, &params).unwrap();
        assert_eq!(result, "INSERT INTO t (name) VALUES ('O''Brien')");
    }
}
